src/pcm/SoxrResampler.cxx

```
@@ -65,6 +85,104 @@ soxr_quality_name(unsigned long recipe) noexcept
+soxr_parse_quality_advanced(const char *copt, struct soxr *r)  noexcept
+{
+	char *opt, *recipe = NULL, *flags = NULL;
+	char *atten = NULL;
+	char *precision = NULL, *passband_end = NULL, *stopband_begin = NULL, *phase_response = NULL;
+
+	opt = xstrdup(copt);
+
+	r->max_rate = false;
+	r->exception = false;
+
+	if (opt) {
+		recipe = quality_advanced_nextparam(opt, ':');
+		flags = quality_advanced_nextparam(NULL, ':');
+		atten = quality_advanced_nextparam(NULL, ':');
+		precision = quality_advanced_nextparam(NULL, ':');
+		passband_end = quality_advanced_nextparam(NULL, ':');
+		stopband_begin = quality_advanced_nextparam(NULL, ':');
+		phase_response = quality_advanced_nextparam(NULL, ':');
+	}
+
+	/* default to HQ (20 bit) if not user specified */
+	r->q_recipe = SOXR_HQ;
+	r->q_flags = 0;
+	/* default to 1db of attenuation if not user specified */
+	r->scale = pow(10, -1.0 / 20);
+	/* override recipe derived values with user specified values */
+	r->q_precision = 0;
+	r->q_passband_end = 0;
+	r->q_stopband_begin = 0;
+	r->q_phase_response = -1;
+
+	if (recipe && recipe[0] != '\0') {
+		if (strchr(recipe, 'v')) r->q_recipe = SOXR_VHQ;
+		if (strchr(recipe, 'h')) r->q_recipe = SOXR_HQ;
+		if (strchr(recipe, 'm')) r->q_recipe = SOXR_MQ;
+		if (strchr(recipe, 'l')) r->q_recipe = SOXR_LQ;
+		if (strchr(recipe, 'q')) r->q_recipe = SOXR_QQ;
+		if (strchr(recipe, 'L')) r->q_recipe |= SOXR_LINEAR_PHASE;
+		if (strchr(recipe, 'I')) r->q_recipe |= SOXR_INTERMEDIATE_PHASE;
+		if (strchr(recipe, 'M')) r->q_recipe |= SOXR_MINIMUM_PHASE;
+		if (strchr(recipe, 's')) r->q_recipe |= SOXR_STEEP_FILTER;
+		/* IGNORED: X = async resampling to max_rate */
+		if (strchr(recipe, 'X')) r->max_rate = true;
+		/* IGNORED: E = exception, only resample if native rate is not */
+		if (strchr(recipe, 'E')) r->exception = true;
+	}
+
+	if (flags)
+		r->q_flags = strtoul(flags, 0, 16);
+
+	if (atten) {
+		double scale = pow(10, -atof(atten) / 20);
+		if (scale > 0 && scale <= 1.0)
```

Bad error check. It silently ignores invalid values.